Modules/atexitmodule.c

```diff
@@ -12,6 +12,16 @@
 #include "pycore_interp.h"        // PyInterpreterState.atexit
 #include "pycore_pystate.h"       // _PyInterpreterState_GET
 
+#ifdef Py_GIL_DISABLED
+/* Note: anything declared as static in this file assumes the lock is held
```

Comment:
Here you should maybe add a macro that can assert that conditon on debug builds.